### PR TITLE
Matplotlib-based circuit drawer

### LINKED ISSUE #311: Implement the Circuit Graph plotter in Python
Currently we have a feature that [plots the circuits to .png](https://github.com/QISKit/qiskit-sdk-py/blob/dfc81048662a9473daf1eb0ad8988ea69ee02a59/qiskit/tools/visualization.py#L686), implemented in #295 (closing #235).

Ideally, it would be easier for the end-user to be able to use that feature if it was implemented in pure Python, as it currently depends on some "heavy" non-Python software (pdflatex, mostly) that might not be available or installable under all environments.

### PR DESCRIPTION
**Pull request by t-imamichi**
### Summary

We implemented a matplotlib-based circuit drawer, which is a drop-in replacement of `latex_drawer`.

### Details and comments

`matplotlib_drawer` supports all functionality as `latex_drawer`.
Users can customize the style of the drawing.
We will write the details in a notebook of qiskit-tutorial.

Example (test.py):
```
from math import pi
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
from qiskit.tools.matplotlibdrawer import matplotlib_drawer

qr = QuantumRegister(3, 'q')
cr = ClassicalRegister(3, 'c')
circuit = QuantumCircuit(qr, cr)
circuit.x(qr[0])
circuit.y(qr[0])
circuit.z(qr[0])
circuit.barrier(qr[0])
circuit.barrier(qr[1])
circuit.barrier(qr[2])
circuit.h(qr[0])
circuit.s(qr[0])
circuit.sdg(qr[0])
circuit.t(qr[0])
circuit.tdg(qr[0])
circuit.iden(qr[0])
circuit.reset(qr[0])
circuit.rx(pi, qr[0])
circuit.ry(pi, qr[0])
circuit.rz(pi, qr[0])
circuit.u0(pi, qr[0])
circuit.u1(pi, qr[0])
circuit.u2(pi, pi, qr[0])
circuit.u3(pi, pi, pi, qr[0])
circuit.swap(qr[0], qr[1])
circuit.cx(qr[0], qr[1])
circuit.cy(qr[0], qr[1])
circuit.cz(qr[0], qr[1])
circuit.ch(qr[0], qr[1])
circuit.cu1(pi, qr[0], qr[1])
circuit.cu3(pi, pi, pi, qr[0], qr[1])
circuit.crz(pi, qr[0], qr[1])
circuit.ccx(qr[0], qr[1], qr[2])
circuit.cswap(qr[0], qr[1], qr[2])
circuit.measure(qr, cr)
matplotlib_drawer(circuit, filename='tmp.pdf') # without barrier
matplotlib_drawer(circuit, filename='tmp2.pdf', style={'plotbarrier': True}) # with barrier
```
Result:
(without barrier, same style as latex_drawer)
image

(with barrier)
image

---
Ali: this resolves #311

### CODE REVIEW

**Comment by ajavadia** [comment]
@t-imamichi thanks for the nice contribution. can you please:
- update the CHANGELOG file with a one-line description of this PR? In the UNRELEASED portion.
- make sure the heading of all new files you add matches the heading of existing files in QISKit. This includes the license information. We also don't add author information to individual files, but rather place it in the [Contributors.md](https://github.com/QISKit/qiskit-core/blob/master/CONTRIBUTORS.md) file.

**Comment by t-imamichi** [comment]
Thank you for your comments. We will revise our codes soon.

File: CONTRIBUTORS.md [hunk 1]
```diff
@@ -27,6 +27,8 @@ involved in the project:
 * John Smolin
 * Erick Winston
 * Chris Wood
+* Takashi Imamichi
```
Comment (ajavadia) [inline CONTRIBUTORS.md]: can you insert these in alphabetical order?

File: tools/qcdrawer.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2017, IBM.
```
Comment (ajavadia) [inline tools/qcdrawer.py]: can you make sure all files you add have copyright 2018?

**Comment by ajavadia** [comment]
Thanks. 
@nkanazawa1989 could you please sign the CLA so we can accept your contribution?

@t-imamichi can you please provide a description of what these 2 files are, or where they are used?
1. `composer.json`
2. `qcdrawer.py`

File: CONTRIBUTORS.md (hunk 1 above)
Comment (t-imamichi) [reply CONTRIBUTORS.md]: Yes, I will do it.

File: tools/qcdrawer.py (hunk 2 above)
Comment (t-imamichi) [reply tools/qcdrawer.py]: Yes, I will do it.

**Comment by t-imamichi** [comment]
They are optional tools. Not mandatory for `matplotlib_drawer.py`.

- `qcdrawer.py` is a stand-alone script to draw a qasm circuit to a png, pdf, or svg file.
- `composer.json` is a style file of drawing, which is based on IBM Q Experience color setting.

If you do not recommend to include them in qiskit-core, it is OK. We will remove them.

File: tools/qcdrawer.py (hunk 2 above)
Comment (t-imamichi) [reply tools/qcdrawer.py]: done

File: CONTRIBUTORS.md (hunk 1 above)
Comment (t-imamichi) [reply CONTRIBUTORS.md]: done

**Comment by t-imamichi** [comment]
FYI: We are preparing a notebook about `matplotlib_drawer` at qiskit-tutorial.
https://github.com/t-imamichi/qiskit-tutorial/blob/drawer/reference/tools/matplotlib_drawer.ipynb

**Comment by t-imamichi** [comment]
@ajavadia We both signed the CLA. What do you think about `qcdrawer.py` and `composer.json`?

**Comment by ajavadia** [comment]
@t-imamichi: I just pushed a commit to your branch which splits the visualization files into 3 parts:
state visualization, circuit visualization and counts visualization.
I put both the matplotlib drawer and the latex drawer under circuit visualization. 

Can you please remove the standalone scripts? The idea is to keep qiskit clean and encourage users to use qiskit components directly. In this case, for example, it is really easy to use `qiskit.load_qasm_file()` and run the drawer on it.
The QX color scheme might be cool to have, but it would be nice if you could include it inside `matplotlib_drawer` and allow choosing it with a flag. For example I see a `QCStyle` class in your code that defines colors. Can you include the QX color schemes there? Of course, this is not urgent, and you can add it as a new feature after this PR is merged.

**Comment by t-imamichi** [comment]
Thanks. I am working on QX color scheme in QCStyle. By the way, I feel `_circuit_visualization.py` is very long. How about splitting it into `_circuit_visualization.py`, `_latex_drawer.py`, and `_matplotlib_drawer.py`?

**Comment by t-imamichi** [comment]
I tried to split `_circuit_visualization.py` in the following branch. https://github.com/t-imamichi/qiskit-core/tree/split_drawer

**Comment by ajavadia** [comment]
I did some refactoring on this. Specifically, we now have 2 methods for creating quantum circuit images:
`latex_circuit_drawer`
`matplotlib_circuit_drawer`
and a wrapper around both of these:
`circuit_drawer`
which tries the 1st one, and in case of failure, falls back on 2nd one.

@delapuente is now going through to clean up the code more, and add some tests. Then we can merge.

**Comment by t-imamichi** [comment]
I notice that `latex_circuit_drawer` is optimized for scale=0.7 and `matplotlib_circuit_drawer` is optimized for scale=1.0. I will change a parameter for `matplotlib_circuit_drawer` to be optimized with scale=0.7

**Comment by t-imamichi** [comment]
@delapuente This implementation always create PNG image because it uses 'circuit.png'.
https://github.com/t-imamichi/qiskit-core/blob/matplotlib_drawer/qiskit/tools/visualization/_circuit_visualization.py#L1541

Matplotlib supports PDF, SVG, PNG image according to the extension of file name. Moreover, the file name is not specified, matplotlib visualizes a figure inline of jupyter notebook by specifying `%matplotlib inline`. So I propose to revert the implementation as in our version.
See our notebook how it works in jupyter notebook in the previous version.
https://github.com/t-imamichi/qiskit-tutorial/blob/drawer/reference/tools/matplotlib_drawer.ipynb

**Comment by t-imamichi** [comment]
@delapuente I changed `matplotlib_circuit_drawer` so that it returns an image object when matplotlib is not inline mode and it generates a PNG figure because `Image` does not support PDF and SVG format. Does it make sense?
Please take a look at the notebook to see how it works.
https://github.com/t-imamichi/qiskit-tutorial/blob/drawer/reference/tools/matplotlib_circuit_drawer.ipynb

**Comment by delapuente** [comment]
Agree with almost all the changes but I have two objections:
* Unless it prevents the Jupyter notebook use-case from working, I prefer to keep a consistent API and return an Image even if unnecessary (let the user ignore the result).
* It is a little bit strange that you specify `circuit_drawer(filename='my_circuit.pdf')`, the Latex drawer run, and write a PNG with the name `my_circuit.pdf`. Isn't it?

**Comment by t-imamichi** [comment]
I agree with the importance of API consistency. How about returning an empty image object when matplotlib inline mode?
When `latex_circuit_drawer` runs, it generates PNG file to obtain the image object to show in a notebook. `latex_circuit_drawer` can save only raster images because of PIL. However, matplotlib is more flexible than PIL and can save vector images. I don't want to disable vector image of matplotlib.

**Comment by t-imamichi** [comment]
The current version of `matplotlib_circuit_drawer` draws a figure twice if inline mode is on. Inline mode is important because it supports svg format, which looks much better than png format.
image

**Comment by t-imamichi** [comment]
I changed it a bit so that it returns an empty image object when matplotlib inline mode is on. It keeps both the API consistency and appearance of the notebook in both cases where inline is on and off.

**Comment by delapuente** [comment]
@t-imamichi I see your point and now I understand why you returned `None`, it draws twice because Jupyter renders the value of an expression automatically and **also** matplot is inlining the results becuase of the directive. We should come up with a better solution for this but, given the urgency, it's fine.

I've open a follow-up to discuss and implement it properly in #612

**Comment by t-imamichi** [comment]
Thank you. It makes sense. Let's continue to improve it there. As my final change, I optimize the font size for the non-inline mode.

**Pull request merged**